addon/-private/system/store.js

```diff
@@ -800,63 +789,67 @@ Store = Service.extend({
     }
 
     this._pendingFetch.forEach(this._flushPendingFetchForType, this);
-    this._pendingFetch = Map.create();
+    this._pendingFetch.clear();
   },
 
-  _flushPendingFetchForType(pendingFetchItems, typeClass) {
-    var store = this;
-    var adapter = store.adapterFor(typeClass.modelName);
-    var shouldCoalesce = !!adapter.findMany && adapter.coalesceFindRequests;
-    var records = Ember.A(pendingFetchItems).mapBy('record');
+  _flushPendingFetchForType(pendingFetchItems, modelClass) {
+    let store = this;
+    let adapter = store.adapterFor(modelClass.modelName);
+    let shouldCoalesce = !!adapter.findMany && adapter.coalesceFindRequests;
+    let totalItems = pendingFetchItems.length;
+    let internalModels = new Array(totalItems);
+    let seeking = new EmptyObject();
+
+    for (let i = 0; i < totalItems; i++) {
+      internalModels[i] = pendingFetchItems[i].internalModel;
+      seeking[internalModels[i].id] = pendingFetchItems[i];
+    }
 
     function _fetchRecord(recordResolverPair) {
-      recordResolverPair.resolver.resolve(store.fetchRecord(recordResolverPair.record, recordResolverPair.options)); // TODO adapter options
+      recordResolverPair.resolver.resolve(store._fetchRecord(recordResolverPair.internalModel, recordResolverPair.options)); // TODO adapter options
     }
 
-    function resolveFoundRecords(records) {
-      records.forEach((record) => {
-        var pair = Ember.A(pendingFetchItems).findBy('record', record);
+    function handleFoundRecords(foundInternalModels, expectedInternalModels) {
+      // resolve found records
+      let found = new EmptyObject();
+      for (let i = 0, l = foundInternalModels.length; i < l; i++) {
+        let internalModel = foundInternalModels[i];
+        let pair = seeking[internalModel.id];
+        found[internalModel.id] = internalModel;
+
         if (pair) {
-          var resolver = pair.resolver;
-          resolver.resolve(record);
+          let resolver = pair.resolver;
+          resolver.resolve(internalModel);
         }
-      });
-      return records;
-    }
+      }
 
-    function makeMissingRecordsRejector(requestedRecords) {
-      return function rejectMissingRecords(resolvedRecords) {
-        resolvedRecords = Ember.A(resolvedRecords);
-        var missingRecords = requestedRecords.reject((record) => resolvedRecords.includes(record));
-        if (missingRecords.length) {
-          warn('Ember Data expected to find records with the following ids in the adapter response but they were missing: ' + Ember.inspect(Ember.A(missingRecords).mapBy('id')), false, {
-            id: 'ds.store.missing-records-from-adapter'
-          });
+      // reject missing records
+      let missingRecords = [];
+      for (let i = 0, l = expectedInternalModels.length; i < l; i++) {
+        if (!found[expectedInternalModels[i].id]) {
+          missingRecords.push(expectedInternalModels[i]);
```

Comment:
same argument as above, I always go back and forth on this. Saving one index access vs one assignment. Perf wise it's likely ignorable, but readability is a bit better.